### PR TITLE
Implement compile_restricted_function

### PR DESCRIPTION
**Pull request by loechel**

### CODE REVIEW

**Comment by dwt** [comment]
please don't merge, still WIP

**Comment by dwt** [comment]
LGTM - my work on the compile function is finished and seems to work with Product.PageTemplate (at least as far as the compiling of scripts is concerned)

File: .coveragerc [hunk 1]
```diff
@@ -2,10 +2,15 @@
 branch = True
 source = RestrictedPython
 omit =
+    bootstrap.py
```
Comment (icemac) [inline .coveragerc]: This seems to be not needed here as it is outside the directories configured for the coverage.

File: .coveragerc [hunk 2]
```diff
@@ -2,10 +2,15 @@
...
+    tests/*.py
```
Comment (icemac) [inline .coveragerc]: Tests should be covered, too.

File: .coveragerc [hunk 3]
```diff
@@ -2,10 +2,15 @@
...
     src/RestrictedPython/tests
     src/RestrictedPython/tests/*.py
-    tests/*.py
-    bootstrap.py
+    # Explicit exclude old implementation for the moment.
+    # should be removed when those files have gone.
+    src/RestrictedPython/SelectCompiler.py
+    src/RestrictedPython/MutatingWalker.py
+    src/RestrictedPython/RestrictionMutator.py
```
Comment (icemac) [inline .coveragerc]: Why? They are tested, too using the old tests.

File: setup.cfg [hunk 4]
```diff
@@ -34,10 +34,22 @@ force_alphabetical_sort = True
...
+    src/RestrictedPython/SelectCompiler.py,
+
+ignore =
+    N801,
+    N802,
+    N803,
+    N805,
+    N806,
+    N812,
+    E301
```
Comment (icemac) [inline setup.cfg]: Really E301?

File: src/RestrictedPython/compile.py [hunk 5]
```diff
@@ -107,32 +112,87 @@ def compile_restricted_single(
 
 
 def compile_restricted_function(
-        p,
+        p,  # parameters
         body,
         name,
         filename='<string>',
-        globalize=None,
+        globalize=None,  # List of globals (e.g. )
```
Comment (icemac) [inline src/RestrictedPython/compile.py]: e. g. what?

File: src/RestrictedPython/compile.py [hunk 6]
```diff
@@ -107,32 +112,87 @@ def compile_restricted_single(
...
+    >>> safe_globals = safe_globals.copy()
+    >>> safe_globals.update(my_call_specific_global_bindings)
+    >>> import types
+    >>> new_function = types.FunctionType(compiled_function.func_code, \
+                                          safe_globals, \
+                                          '<function_name>', \
+                                          compiled_function.func_defaults or \
+                                          () \
+                                          )
+    >>> result = new_function(*[], **{})
```
Comment (icemac) [inline src/RestrictedPython/compile.py]: Is this test code actually run? Should it be moved to the Sphinx documentation where we could test it?

File: tests/__init__.py [hunk 7]
```diff
@@ -44,10 +66,21 @@ def _eval(source, glb=None):
...
+e_function = ('e_function', [_function(RestrictedPython.compile.compile_restricted_function)])  # NOQA: E501
+
+c_single = ('c_single', [RestrictedPython.compile.compile_restricted_single])
+e_single = ('e_single', [_single(RestrictedPython.compile.compile_restricted_single)])  # NOQA: E501
 
 if IS_PY2:
     from RestrictedPython import RCompile
     c_exec[1].append(RCompile.compile_restricted_exec)
     c_eval[1].append(RCompile.compile_restricted_eval)
+    # c_single[1].append(lambda source, filename: RCompile.compile_restricted(source, filename, 'single'))  # NOQA: E501
```
Comment (icemac) [inline tests/__init__.py]: Why has this code a comment sign?

File: tests/__init__.py [hunk 8]
```diff
@@ -44,10 +66,21 @@ def _eval(source, glb=None):
...
+    c_function[1].append(RCompile.compile_restricted_function)
+
     e_exec[1].append(_exec(RCompile.compile_restricted_exec))
     e_eval[1].append(_eval(RCompile.compile_restricted_eval))
+    # e_single[1].append(_single(lambda source, filename: RCompile.compile_restricted(source, filename, 'single')))  # NOQA: E501
```
Comment (icemac) [inline tests/__init__.py]: Ditto.

File: tests/test_compile.py [hunk 9]
```diff
@@ -155,3 +158,47 @@ def test_compile__compile_restricted_eval__used_names(c_eval):
     assert result.errors == ()
     assert result.warnings == []
     assert result.used_names == {'a': True, 'b': True, 'x': True, 'func': True}
+
+
+@pytest.mark.parametrize(*c_single)
+def test_compile__compile_restricted_csingle(c_single):
+    """It compiles code as an Expression."""
```
Comment (icemac) [inline tests/test_compile.py]: This doc string does not seem to match the function body.

File: tests/test_compile.py [hunk 10]
```diff
@@ -155,3 +158,47 @@ def test_compile__compile_restricted_eval__used_names(c_eval):
...
+def test_compile_restricted():
+    with pytest.warns(SyntaxWarning) as record:
+        result = compile_restricted(PRINT_EXAMPLE, '<string>', 'exec')
+        assert isinstance(result, types.CodeType)
+        assert len(record) == 2
+        assert record[0].message.args[0] == \
+            'Line 3: Print statement is deprecated ' \
+            'and not avaliable anymore in Python 3.'
+        assert record[1].message.args[0] == \
+            "Line 2: Prints, but never reads 'printed' variable."
```
Comment (icemac) [inline tests/test_compile.py]: There is already `test_print_stmt.py` where the test for the warning could be added once in the `test_print_stmt__simple_prints` function.

File: tests/test_compile.py [hunk 11]
```diff
@@ -155,3 +158,47 @@ def test_compile__compile_restricted_eval__used_names(c_eval):
...
+EVAL_EXAMPLE = """
+def a():
+    eval('2 + 2')
+"""
+
+
+def test_compile_restricted_eval():
+    with pytest.raises(SyntaxError,
+                       message="Line 3: Eval calls are not allowed."):
+        compile_restricted(EVAL_EXAMPLE, '<string>', 'exec')
```
Comment (icemac) [inline tests/test_compile.py]: This is already tested in `test_transformer.test_transformer__RestrictingNodeTransformer__visit_Call__3`. Or am I missing a difference here between those two tests?

File: tests/test_compile_restricted_function.py [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+from RestrictedPython import safe_builtins
+from tests import c_function
+from types import FunctionType
+
+import pytest
+
+
+@pytest.mark.parametrize(*c_function)
+def test_compile_restricted_function(c_function):
+    p = ''
```
Comment (icemac) [inline tests/test_compile_restricted_function.py]: Might it be clearer to use an empty  tuple here?

File: tests/test_compile_restricted_function.py [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+    safe_locals = {}
+    exec(result.code, safe_globals, safe_locals)
+    hello_world = safe_locals['hello_world']
+    assert type(hello_world) == FunctionType
+    assert hello_world() == 'Hello World!\n'
+
+
+@pytest.mark.parametrize(*c_function)
+def test_compile_restricted_function_func_wrapped(c_function):
+    p = ''
```
Comment (icemac) [inline tests/test_compile_restricted_function.py]: Ditto.

File: tests/test_compile_restricted_function.py [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+    func = FunctionType(result.code, safe_globals)
+    func()
+    assert 'hello_world' in safe_globals
+    hello_world = safe_globals['hello_world']
+    assert hello_world() == 'Hello World!\n'
+
+
+@pytest.mark.parametrize(*c_function)
+def test_compile_restricted_function_with_arguments(c_function):
+    p = 'input'
```
Comment (icemac) [inline tests/test_compile_restricted_function.py]: Hm, this is a string, who do I use multiple parameters?

File: tox.ini [hunk 15]
```diff
@@ -21,7 +21,8 @@ extras =
     develop
     test
 commands =
-    py.test --cov=src --cov-report=xml --html=report-{envname}.html --self-contained-html {posargs}
+    pytest --cov=src --cov-report=xml --html=report-{envname}.html --self-contained-html {posargs}
+    #pytest --doctest-modules src/RestrictedPython/compile.py {posargs}
```
Comment (icemac) [inline tox.ini]: Why is there the `#`?

File: tests/test_compile_restricted_function.py (hunk 12 above)
Comment (loechel) [inline tests/test_compile_restricted_function.py]: need to be a string

File: tests/test_compile_restricted_function.py (hunk 13 above)
Comment (loechel) [inline tests/test_compile_restricted_function.py]: ditto

File: tests/test_compile.py (hunk 10 above)
Comment (loechel) [inline tests/test_compile.py]: see new docstring

**Pull request merged**